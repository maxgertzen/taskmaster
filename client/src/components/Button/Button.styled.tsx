import styled from '@emotion/styled';

export const StyledButton = styled.button<{
  variant: 'primary' | 'secondary' | 'outline' | 'danger';
}>(({ theme, variant }) => ({
  padding: theme.spacing(1, 2),
  backgroundColor: theme.colors[variant === 'danger' ? 'outline' : variant],
  color:
    variant === 'outline'
      ? theme.colors.text
      : variant === 'danger'
        ? theme.colors.danger
        : theme.colors.background,
  fontWeight: 600,
  textTransform: 'uppercase',
  border: 'none',
  borderRadius: theme.borders.radius,
  cursor: 'pointer',
  transition: 'background-color 0.3s',
  '&:hover': {
    backgroundColor: theme.colors.secondary,
  },
  ...((variant === 'outline' || variant === 'danger') && {
    '&:hover': {
      backgroundColor:
        variant === 'danger' ? theme.colors.danger : theme.colors.accent,
      color: theme.colors.background,
    },
    border: theme.borders.main(
      variant === 'danger' ? theme.colors.danger : theme.colors.text
    ),
  }),
}));
