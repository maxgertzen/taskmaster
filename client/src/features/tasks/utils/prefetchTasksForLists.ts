import { QueryClient } from '@tanstack/react-query';

import { QUERY_KEYS } from '@/shared/api/query-keys';
import { STALE_TIME } from '@/shared/constants/staleTime';

import { fetchTasks } from '../api';

export const prefetchTasksForLists = (
  listIds: string[],
  queryClient: QueryClient,
  token: string | null
) => {
  if (!token || !listIds.length) return;

  listIds.forEach((listId) => {
    queryClient.prefetchQuery({
      queryKey: QUERY_KEYS.tasks({ listId }),
      queryFn: fetchTasks(token),
      staleTime: STALE_TIME,
    });
  });
};
