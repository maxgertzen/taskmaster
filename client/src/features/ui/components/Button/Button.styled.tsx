import styled from '@emotion/styled';

import PrimaryEmboss from '@/shared/assets/images/primary-emboss.png';
import PrimaryOutline from '@/shared/assets/images/primary-empty.png';
import PrimaryFilled from '@/shared/assets/images/primary-filled.png';
import SuccessEmpty from '@/shared/assets/images/success-empty.png';
import SuccessFilled from '@/shared/assets/images/success-filled.png';
import { ButtonType } from '@/shared/types/ui';

const buttonMap: Record<string, string> = {
  primary: PrimaryOutline,
  primaryHover: PrimaryFilled,
  secondary: PrimaryEmboss,
  secondaryHover: PrimaryFilled,
  success: SuccessEmpty,
  successHover: SuccessFilled,
};
export const StyledButton = styled.button(({ theme }) => ({
  fontSize: theme.typography.body.fontSize,
  textTransform: 'uppercase',
  border: 'none',
  cursor: 'pointer',
  backgroundColor: 'transparent',
  height: theme.spacing(3),
  width: '100%',
  padding: theme.spacing(2),
  whiteSpace: 'normal',
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',

  [`@media (max-width: ${theme.breakpoints.sm})`]: {
    padding: theme.spacing(1, 2),
    fontSize: theme.spacing(1),
  },
}));

export const ButtonWrapper = styled.div<{
  variant: ButtonType;
  isActive: boolean;
  isDarkTheme: boolean;
}>(({ theme, variant, isActive, isDarkTheme }) => ({
  position: 'relative',
  minWidth: 'fit-content',
  color: isDarkTheme ? theme.colors.background : theme.colors.text,
  backgroundPosition: 'center',
  backgroundSize: '100% 100%',
  backgroundImage: isActive
    ? `url(${buttonMap[`${variant}Hover`]})`
    : `url(${buttonMap[variant]})`,
  backgroundRepeat: 'no-repeat',
  '&:hover': {
    color: theme.colors.background,
    backgroundImage: `url(${buttonMap[`${variant}Hover`]})`,
  },
  pointerEvents: isActive ? 'none' : 'auto',
  transition: 'background-image 0.3s',
}));
