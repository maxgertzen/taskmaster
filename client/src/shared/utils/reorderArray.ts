export const reorderArray = <T>(
  array: T[],
  fromIndex: number,
  toIndex: number
): T[] => {
  const reordered = [...array];
  const [movedItem] = reordered.splice(fromIndex, 1);
  reordered.splice(toIndex, 0, movedItem);
  return reordered;
};
