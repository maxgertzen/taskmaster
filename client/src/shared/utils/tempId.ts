export const generateTempId = (prefix: string = 'temp') =>
  `${prefix}-${Date.now()}`;
