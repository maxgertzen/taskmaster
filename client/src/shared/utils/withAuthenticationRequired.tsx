import { withAuthenticationRequired as realWithAuthenticationRequired } from '@auth0/auth0-react';

import { IS_AUTH0_DISABLED } from '@/constants';
import { mockWithAuthenticationRequired } from '@/mocks';

const isMock = IS_AUTH0_DISABLED === 'true';

export const withAuthenticationRequired = isMock
  ? mockWithAuthenticationRequired
  : realWithAuthenticationRequired;
