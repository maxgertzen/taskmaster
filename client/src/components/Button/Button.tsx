import { FC } from 'react';

import { useThemeStore } from '../../store/themeStore';
import { ButtonType } from '../../types/shared';

import { ButtonWrapper, StyledButton } from './Button.styled';

interface ButtonProps {
  children: string;
  onClick?: () => void;
  variant?: ButtonType;
  disabled?: boolean;
  isActive?: boolean;
}

export const Button: FC<ButtonProps> = ({
  children,
  onClick,
  disabled = false,
  variant = 'primary',
  isActive = false,
}) => {
  const theme = useThemeStore((state) => state.theme);
  return (
    <ButtonWrapper
      variant={variant}
      isActive={isActive}
      isDarkTheme={theme === 'dark'}
    >
      <StyledButton disabled={disabled} onClick={onClick}>
        {children}
      </StyledButton>
    </ButtonWrapper>
  );
};
