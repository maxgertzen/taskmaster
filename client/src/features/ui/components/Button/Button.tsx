import { FC } from 'react';

import { useThemeStore } from '@/shared/store/themeStore';
import { ButtonType } from '@/shared/types/ui';

import { ButtonWrapper, StyledButton } from './Button.styled';

interface ButtonProps {
  children: string;
  onClick?: () => void;
  variant?: ButtonType;
  disabled?: boolean;
  isActive?: boolean;
}

export const Button: FC<ButtonProps> = ({
  children,
  onClick,
  disabled = false,
  variant = 'primary',
  isActive = false,
}) => {
  const theme = useThemeStore((state) => state.theme);
  return (
    <ButtonWrapper
      variant={variant}
      isActive={isActive}
      isDarkTheme={theme === 'dark'}
    >
      <StyledButton disabled={disabled} onClick={onClick}>
        {children}
      </StyledButton>
    </ButtonWrapper>
  );
};
