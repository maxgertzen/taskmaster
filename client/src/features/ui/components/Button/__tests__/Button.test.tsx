import { describe, it, expect, vi } from 'vitest';

import { render, screen, within } from '@/tests/utils';

import { Button } from '../Button';

describe('Button', () => {
  it('renders correctly in light theme', () => {
    const { container } = render(<Button>Click me</Button>, { theme: 'light' });
    expect(container).toMatchSnapshot();
  });

  it('renders correctly in dark theme', () => {
    const { container } = render(<Button>Click me</Button>, { theme: 'dark' });
    expect(container).toMatchSnapshot();
  });

  it('renders the children', () => {
    render(<Button>Click me</Button>);
    const button = screen.getByTestId('button');
    expect(button).toBeDefined();

    const text = within(button).getByText(/click me/i);
    expect(text.textContent).toBe('Click me');
  });

  it('calls the onClick function when clicked', async () => {
    const handleClick = vi.fn();
    const { user } = render(<Button onClick={handleClick}>Click me</Button>);

    const button = screen.getByTestId('button');
    await user.click(button);

    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  it('is disabled when disabled prop is true', async () => {
    const handleClick = vi.fn();
    const { user } = render(
      <Button disabled onClick={handleClick}>
        Click me
      </Button>
    );

    const button = screen.getByTestId('button');
    expect(button).toBeDisabled();

    await user.click(button);
    expect(handleClick).not.toHaveBeenCalled();
  });
});
