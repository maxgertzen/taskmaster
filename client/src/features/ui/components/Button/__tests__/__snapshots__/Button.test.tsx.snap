// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > renders correctly 1`] = `
<div>
  <div
    class="css-aud1u0"
  >
    <button
      class="css-1p3b2wb"
      data-testid="button"
    >
      Click me
    </button>
  </div>
</div>
`;

exports[`Button > renders correctly in dark theme 1`] = `
<div>
  <div
    class="css-pl2bwv"
  >
    <button
      class="css-1p3b2wb"
      data-testid="button"
    >
      Click me
    </button>
  </div>
</div>
`;

exports[`Button > renders correctly in light theme 1`] = `
<div>
  <div
    class="css-aud1u0"
  >
    <button
      class="css-1p3b2wb"
      data-testid="button"
    >
      Click me
    </button>
  </div>
</div>
`;
